### PR TITLE
Add CommandBatchPreparerDependencies and StateManagerDependencies

### PR DESCRIPTION
**Pull request by AndriySvyryd**
These will allow to create `CommandBatchPreparer` and `StateManager` instances on demand.

### CODE REVIEW

File: src/EFCore.Relational/Update/Internal/CommandBatchPreparerDependencies.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public ILoggingOptions LoggingOptions { get; }
+
+        /// <summary>
+        ///     Clones this dependency parameter object with one service replaced.
+        /// </summary>
+        /// <param name="modificationCommandBatchFactory"> A replacement for the current dependency of this type. </param>
+        /// <returns> A new parameter object with the given service replaced. </returns>
+        public CommandBatchPreparerDependencies With([NotNull] IModificationCommandBatchFactory modificationCommandBatchFactory)
```
Comment (Meir017) [inline src/EFCore.Relational/Update/Internal/CommandBatchPreparerDependencies.cs]: any reason not to make this method an extension method? (and the other `With` methods)

Comment (AndriySvyryd) [reply src/EFCore.Relational/Update/Internal/CommandBatchPreparerDependencies.cs]: We don't use extension methods unless there's a good reason to. This improves cohesion.

Comment (Meir017) [reply src/EFCore.Relational/Update/Internal/CommandBatchPreparerDependencies.cs]: These methods seem to me like utility methods and therefore they should not be inside the class, they don't use any private members or anything like that

Comment (AndriySvyryd) [reply src/EFCore.Relational/Update/Internal/CommandBatchPreparerDependencies.cs]: The whole purpose of this class is to hold and manipulate dependencies. The `With` methods are part of its responsibilities and contract.

**Pull request merged**